### PR TITLE
feat: bash tool change attribution via filesystem snapshots, codex fixes

### LINKED ISSUE #150: Preserve attributions through intra-commit file renames/moves

### LINKED ISSUE #756: Bash tool file changes not attributed to AI — only Write|Edit|MultiEdit are tracked
Problem         

  When Claude Code modifies files via the Bash tool (e.g. cp, mv, code generators, build scripts), those changes are not attributed to AI. They silently appear as human-authored lines in git-ai blame and stats.

  Root cause

  The Claude Code hook installer (claude_code.rs) sets the PostToolUse matcher to Write|Edit|MultiEdit only. The Bash tool is not included, so no checkpoint is triggered when Bash modifies files.

  Reproduction

  1. In a Claude Code session, ask Claude to copy or generate a file using Bash (e.g. cp foo.rs bar.rs)
  2. Commit the result
  3. Run git-ai show <commit> — prompts is empty {}
  4. Run git-ai blame — all lines show as human-authored

  Expected behavior

  File changes caused by the Bash tool should be attributed to AI, same as Write/Edit.                                                                                                                                                                                                                                                                                                                                    
   
  Suggested fix                                                                                                                                                                                                                                                                                                                                                                                                           
                  
  1. Add Bash to the PostToolUse matcher in claude_code.rs
  2. In ClaudePreset::run (agent_presets.rs), handle tool_name == "Bash" by setting edited_filepaths to None so the checkpoint falls back to scanning dirty files via git status, rather than relying on tool_input.file_path (which Bash doesn't provide)

### PR DESCRIPTION
**Pull request by jwiegley**
## Summary

- **Bash tool change attribution**: Implements a stat-tuple-based filesystem snapshot system that detects which files were created, modified, or deleted when an AI coding agent executes bash commands. This enables accurate AI provenance tracking for tool invocations that bypass the normal edit/write tool path.
- **Multi-agent preset support**: Integrates bash tool checkpoint handling into Claude, Gemini, Droid, ContinueCli, Amp, and OpenCode presets with tool classification filtering (only triggers on bash/shell tool invocations).
- **Async pre-hook content capture**: Daemon watermark system tracks per-file and per-worktree mtimes so the bash pre-hook can identify and capture human-authored changes without blocking the tool call. Captured checkpoints are written to disk and consumed asynchronously by the daemon.
- **Two-tier path filtering**: Git-tracked files are always included in change detection; new untracked files are filtered against frozen `.gitignore` rules to avoid noise from build artifacts.
- **Rename/move attribution preservation**: Properly tracks both source and destination paths during file renames and moves, ensuring AI provenance transfers correctly through intra-commit file reorganization.
- **Hook catch-all migration**: Claude Code, Droid/Factory, and Gemini hook registrations migrated from specific-tool matchers (`"Write|Edit|MultiEdit"`, `"^(Edit|Write|Create|ApplyPatch)$"`, `"write_file|replace"`) to the catch-all `"*"` matcher so every tool call — including bash — triggers pre/post checkpoints. Full TDD migration with S1–S13/U1–U4/C1–C3 test matrix per agent.
- **OpenCode plugin bash support**: OpenCode TypeScript plugin now handles bash tool calls in a dedicated branch (repo resolution from `workdir`/`cwd` without file-path extraction), alongside the existing file-edit tool path.
- **Comprehensive test coverage** (3,100+ lines across 3 test suites):
  - `bash_tool_conformance.rs` — conformance tests validating file mutations, hook semantics, tool classification, gitignore filtering, rename handling, snapshot save/load
  - `bash_tool_benchmark.rs` — performance tests covering snapshot timing, large repo scaling, memory bounds, concurrent session isolation
  - `bash_tool_provenance.rs` — end-to-end tests exercising real bash commands (file creation, modification, deletion, build tools, git operations, pipelines, symlinks, batch ops, archives, edge cases)

### Devin review fixes (latest commit)

- **B1** — `tool_use_id` collision for agents without per-call IDs (Gemini, ContinueCli): sidecar UUID file in `.git/ai/bash_snapshots/` correlates pre/post hooks, preventing snapshot key collisions on sequential bash calls
- **B2** — Deleted files excluded from post-hook captured checkpoint `edited_filepaths` (no post-content to capture; were stored as empty blobs)
- **W1** — `compute_watermarks_from_stat` now uses `symlink_metadata` (lstat) to match snapshot mtime semantics
- **W2** — Cold-start stale-file dedup normalizes git-status paths before `contains()` check (fixes false duplicates on macOS/Windows case-insensitive FS)
- **W3** — Worktree watermark now set for all live Human checkpoints (previously the `is_empty()` guard prevented it from ever firing for bare-CLI checkpoints)
- **W4** — `flake.nix` version bumped to match `Cargo.toml`

Closes #150
Closes #756

## Test plan

- [x] All bash tool tests pass (`cargo test bash_tool`)
- [x] All agent hook migration tests pass — S1–S13, U1–U4, C1–C3 for Claude Code, Droid, Gemini (`cargo test -p git-ai -- claude_code droid gemini`)
- [x] Full test suite passes (`cargo test -p git-ai`)
- [x] `cargo clippy` — zero errors
- [x] `cargo fmt` — no formatting issues
- [x] Rebased on latest `origin/main`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 1]
```diff
@@ -0,0 +1,1074 @@
...
+        if line.starts_with("1 ") || line.starts_with("u ") {
+            // Ordinary or unmerged entry: last space-separated field is the path
+            if let Some(path) = line.split(' ').next_back() {
+                changed_files.push(crate::utils::normalize_to_posix(path));
+            }
+        } else if line.starts_with("2 ") {
+            // Rename/copy: path is last field, plus next NUL-delimited original path
+            if let Some(path) = line.split(' ').next_back() {
+                changed_files.push(crate::utils::normalize_to_posix(path));
+            }
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs:950]: Fixed in 7954b9ac. Switched to `splitn` with the correct field counts for each porcelain v2 entry type (9 for ordinary `1 `, 11 for unmerged `u `, 10 for rename `2 `). Also added two regression tests: `test_git_status_fallback_files_with_spaces` and `test_git_status_fallback_new_untracked_with_spaces`, both passing.

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 2]
```diff
@@ -0,0 +1,1083 @@
...
+        .hidden(false) // Don't skip hidden files
+        .git_ignore(false) // We handle gitignore ourselves (frozen rules)
+        .git_global(false)
+        .git_exclude(false)
+        .filter_entry(|entry| {
+            // Skip .git directory itself
+            let path = entry.path();
+            !path.components().any(|c| c.as_os_str() == ".git")
+        })
+        .build();
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs:676]: Fixed in 73c9a5c9. The `WalkBuilder` now uses `git_ignore(true)`, `git_global(true)`, and `git_exclude(true)` so it prunes ignored directories during traversal rather than visiting all files only to filter post-walk. A second pass over `tracked_files` ensures the Tier 1 guarantee (git-tracked files always appear in snapshots) is maintained. Test coverage added in dc3499af (`test_snapshot_walker_prunes_ignored_directories`).

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 3]
```diff
@@ -0,0 +1,1083 @@
...
+                        debug_log(&format!(
+                            "Warning: failed to parse {}: {}",
+                            nested_gitignore.display(),
+                            err
+                        ));
+                    }
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: Fixed in 73c9a5c9. `collect_gitignores` is now recursive with a depth limit of 10 (`MAX_GITIGNORE_DEPTH`), so `.gitignore` files at any reasonable nesting depth (e.g., `src/generated/.gitignore`) are discovered and included in the frozen gitignore filter. Test coverage added in dc3499af (`test_build_gitignore_deeply_nested_rules`).

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 4]
```diff
@@ -0,0 +1,1099 @@
...
+
+        // Recurse into subdirectories to find nested .gitignore files
+        if let Ok(entries) = fs::read_dir(dir) {
+            for entry in entries.flatten() {
+                let path = entry.path();
+                if path.is_dir() && !path.ends_with(".git") {
+                    collect_gitignores(builder, &path, depth + 1);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: Fixed in 89b6e3c2. `collect_gitignores` now has a 2-second deadline (`Instant::now() + Duration::from_secs(2)`) and skips well-known large ignored directory names (`node_modules`, `target`, `vendor`, `.build`, `__pycache__`, `.venv`, `dist`, `build`) via a `SKIP_DIR_NAMES` constant. This prevents unbounded blocking before the walker's timeout even starts.

File: src/commands/checkpoint_agent/agent_presets.rs [hunk 5]
```diff
@@ -142,14 +173,35 @@ impl AgentCheckpointPreset for ClaudePreset {
...
+        let edited_filepaths = if is_bash_tool {
+            let repo_root = Path::new(cwd);
+            match bash_tool::handle_bash_tool(HookEvent::PostToolUse, repo_root, session_id, "bash")
+            {
+                Ok(BashCheckpointAction::Checkpoint(paths)) => Some(paths),
+                Ok(BashCheckpointAction::NoChanges) => None,
+                Ok(BashCheckpointAction::Fallback) => {
+                    // Fall back to git status
+                    bash_tool::git_status_fallback(Path::new(cwd)).ok()
+                }
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:322]: Fixed in 9e8c7b8d. All 6 preset `Fallback` arms now return `None` directly instead of redundantly calling `git_status_fallback` again. Since `Fallback` is only returned when `git_status_fallback` already failed inside `handle_bash_tool`, retrying would produce the same failure.

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 6]
```diff
@@ -0,0 +1,1126 @@
...
+    let walker = WalkBuilder::new(repo_root)
+        .hidden(false) // Don't skip hidden files
+        .git_ignore(true) // Prune ignored directories during traversal
+        .git_global(true)
+        .git_exclude(true)
+        .filter_entry(|entry| {
+            // Skip .git directory itself
+            let path = entry.path();
+            !path.components().any(|c| c.as_os_str() == ".git")
+        })
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: Fixed in commit 5a8c2024. Changed `filter_entry` to use `entry.file_name() != ".git"` instead of scanning all absolute path components, exactly as suggested. This prevents false exclusions when the repo absolute path itself contains a `.git` component (e.g., git worktrees).

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 7]
```diff
@@ -0,0 +1,1126 @@
...
+/// On `PreToolUse`: takes a pre-snapshot and stores it.
+/// On `PostToolUse`: takes a post-snapshot, diffs against the stored pre-snapshot,
+/// and returns the list of changed files.
+pub fn handle_bash_tool(
+    hook_event: HookEvent,
+    repo_root: &Path,
+    session_id: &str,
+    tool_use_id: &str,
+) -> Result<BashCheckpointAction, GitAiError> {
+    let invocation_key = format!("{}:{}", session_id, tool_use_id);
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: Fixed in commit ea999e68. All six agent presets now extract the actual `tool_use_id` from hook data instead of hardcoding `"bash"`:

- **Claude, Gemini, ContinueCli, Droid** (`agent_presets.rs`): Extract from `hook_data` JSON, checking both `tool_use_id` and `toolUseId` keys
- **Amp** (`amp_preset.rs`): Uses the already-parsed `hook_input.tool_use_id` field
- **OpenCode** (`opencode_preset.rs`): Added `tool_use_id` field with `#[serde(alias = "toolUseId")]` to struct

All presets fall back to `"bash"` if the field is absent, maintaining backward compatibility while ensuring unique snapshot keys (`{session_id}:{tool_use_id}`) for concurrent invocations.

File: src/commands/checkpoint_agent/bash_tool.rs [hunk 8]
```diff
@@ -0,0 +1,1125 @@
...
+                continue;
+            }
+        };
+
+        let abs_path = entry.path();
+
+        // Skip directories themselves (we only stat files)
+        if abs_path.is_dir() {
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: Good catch — fixed in bffbba71. Changed to use `entry.file_type().map(|ft| ft.is_dir()).unwrap_or_else(|| abs_path.is_dir())` which uses lstat semantics from the walker entry, avoiding following symlinks when filtering directory entries. The `unwrap_or_else` fallback to `abs_path.is_dir()` handles the edge case where `file_type()` returns `None`.

File: src/commands/checkpoint_agent/agent_presets.rs [hunk 9]
```diff
@@ -666,144 +1408,82 @@ impl AgentCheckpointPreset for WindsurfPreset {
...
             checkpoint_kind: CheckpointKind::AiAgent,
             transcript: Some(transcript),
-            repo_working_dir: cwd,
-            edited_filepaths: file_path_as_vec,
+            // use default.
+            repo_working_dir: None,
+            edited_filepaths,
             will_edit_filepaths: None,
             dirty_files: None,
         })
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:1063]: Fixed in commit c3369d4d: restored `repo_working_dir: Some(cwd.to_string())` for both GeminiPreset and ContinueCliPreset. The rebase conflict resolution had incorrectly kept the `None` variant.

File: src/commands/checkpoint_agent/agent_presets.rs [hunk 10]
```diff
@@ -883,27 +1569,74 @@ impl AgentCheckpointPreset for ContinueCliPreset {
...
             checkpoint_kind: CheckpointKind::AiAgent,
             transcript: Some(transcript),
-            repo_working_dir: Some(cwd.to_string()),
-            edited_filepaths: file_path_as_vec,
+            // use default.
+            repo_working_dir: None,
+            edited_filepaths,
             will_edit_filepaths: None,
             dirty_files: None,
         })
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:1233]: Fixed in commit c3369d4d: restored `repo_working_dir: Some(cwd.to_string())` for both GeminiPreset and ContinueCliPreset. The rebase conflict resolution had incorrectly kept the `None` variant.

File: src/commands/checkpoint_agent/agent_presets.rs [hunk 11]
```diff
@@ -801,9 +764,156 @@ impl WindsurfPreset {
...
+            return Ok(AgentRunResult {
+                agent_id,
+                agent_metadata: None,
+                checkpoint_kind: CheckpointKind::Human,
+                transcript: None,
+                repo_working_dir: None,
+                edited_filepaths: None,
+                will_edit_filepaths: file_path_as_vec,
+                dirty_files: None,
+            });
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:1014]: Fixed in commit c3369d4d: restored `repo_working_dir: Some(cwd.to_string())` for both GeminiPreset and ContinueCliPreset. The rebase conflict resolution had incorrectly kept the `None` variant.

File: src/commands/checkpoint_agent/agent_presets.rs [hunk 12]
```diff
@@ -883,27 +999,74 @@ impl AgentCheckpointPreset for ContinueCliPreset {
...
                 agent_id,
                 agent_metadata: None,
                 checkpoint_kind: CheckpointKind::Human,
                 transcript: None,
-                repo_working_dir: Some(cwd.to_string()),
+                repo_working_dir: None,
                 edited_filepaths: None,
                 will_edit_filepaths: file_path_as_vec,
                 dirty_files: None,
             });
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:1184]: Fixed in commit c3369d4d: restored `repo_working_dir: Some(cwd.to_string())` for both GeminiPreset and ContinueCliPreset. The rebase conflict resolution had incorrectly kept the `None` variant.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs:950]: ✅ **Resolved**: Fixed by using `splitn` for proper field parsing. Ordinary entries use `splitn(9, ' ')`, unmerged entries use `splitn(11, ' ')`, and rename entries use `splitn(10, ' ')`. See lines 694-706.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs:676]: ✅ **Resolved**: The walker now uses `git_ignore(true)`, `git_global(true)`, and `git_exclude(true)` (lines 410-412), which enables proper gitignore-based directory pruning during traversal. The frozen gitignore from `build_gitignore()` is still used separately in `diff()` for Tier 2 filtering of new files.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: ✅ **Resolved**: `collect_gitignores` now recurses up to `MAX_GITIGNORE_DEPTH = 10` levels deep (line 297), with a 2-second deadline to prevent unbounded traversal. Well-known large directories (node_modules, target, etc.) are skipped via `SKIP_DIR_NAMES`.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: ✅ **Resolved**: `collect_gitignores` now accepts a `deadline: Instant` parameter and checks `Instant::now() > deadline` at the top of each call (line 320). The deadline is set to 2 seconds from start (line 352).

File: src/commands/checkpoint_agent/agent_presets.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/agent_presets.rs:322]: ✅ **Resolved**: The `Fallback` match arm now returns `None` instead of calling `git_status_fallback` again (lines 193-196). The fallback is only called in the `Err(e)` arm where `handle_bash_tool` itself errored, not when it returned `Fallback`.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: ✅ **Resolved**: The `filter_entry` closure now uses `entry.file_name() != ".git"` (line 415) instead of checking absolute path components, which correctly handles repos with `.git` in ancestor directory names.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: ✅ **Resolved**: GeminiPreset now extracts tool_use_id from hook data instead of hardcoding "bash". Each bash tool invocation gets a unique snapshot key.

File: src/commands/checkpoint_agent/bash_tool.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/commands/checkpoint_agent/bash_tool.rs]: ✅ **Resolved**: The snapshot walker now uses `entry.file_type()` (lstat semantics, lines 440-443) instead of `abs_path.is_dir()` to avoid following symlinks. A symlink to a directory is now correctly snapshotted as a symlink entry rather than being skipped.

**Comment by jwiegley** [comment]
* **#842** <a href="https://app.graphite.com/github/pr/git-ai-project/git-ai/842?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#798** <a href="https://app.graphite.com/github/pr/git-ai-project/git-ai/798?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/git-ai-project/git-ai/798?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
